### ISSUE TITLE
Custom provider support

### ISSUE THREAD

**Issue by jyecusch** [nitrictech/roadmap#24]
### summary

Allow developers to build their own implementations of Nitric providers and use them with the Nitric CLI.

### intended outcome

Allows for full custom provider development, where Nitric framework users can take full control of their deployment process or customize an existing provider.

**Comment by geoman2** [comment]
any progress on this, Jye?

**Comment by jyecusch** [comment]
> any progress on this, Jye?

Absolutely, thanks for checking in. The providers previously were baked directly into our CLI, they have now been separated so they can be downloaded and run independently. Since they communicate using gRPC, new providers can now also be written in languages other than Go, which we're hoping is helpful for teams using languages like TypeScript or Python to build custom solutions.

The first cut of the documentation on this functionality is [here](https://nitric.io/docs/reference/providers/custom/building-custom-provider). This documentation could be hard to follow since the Nitric contributors are the only people who've really worked on providers at this stage.

We'd be happy to provide assistance or guidance if anyone wants to start building their own providers or contribute to the existing providers. We'll update the docs in turn to reflect what we learn from providing that help.

@tjholm also built a recent prototype provider that uses Terraform to deploy to AWS. It currently doesn't support all of the Nitric resources and requires a custom step to build containers, but we'd be happy to let you take a look if that would be useful. Otherwise, our existing Pulumi based providers are a good place to start.

**Comment by raksiv** [comment]
@geoman2 do you have a specific resource that you'd like to build a custom provider for? 

This is an area we'd like to put more focus on, so anything you can share would help us refine requirements.

**Comment by geoman2** [comment]
Scaleway and cloudflare integration would be nice to have

**Comment by raksiv** [comment]
@geoman2 thanks for the response, those are good suggestions.

Couple of followups - 
- Is there a project you're working on which requires scaleway or cloudflare? 
- Do you have any interest in co-developing a contribution?
